client/order/orderbook.go

```diff
@@ -0,0 +1,226 @@
+		}
+		ob.sells[order.Rate] = append(ob.sells[order.Rate], order)
+	}
+}
+
+// removeOrder removes an order from the order book. This must be called with
+// the order book mutex held.
+func (ob *OrderBook) removeOrder(order *msgjson.BookOrderNote) error {
+	switch order.Side {
+	case msgjson.BuyOrderNum:
+		ordersAtPrice, ok := ob.buys[order.Rate]
+		if !ok {
+			return fmt.Errorf("no buy orders found at price %v", order.Rate)
+		}
+
+		for idx := len(ordersAtPrice) - 1; idx >= 0; idx-- {
+			if bytes.Equal(ordersAtPrice[idx].OrderID, order.OrderID) {
+				ordersAtPrice = append(ordersAtPrice[:idx], ordersAtPrice[idx+1:]...)
+				ob.buys[order.Rate] = ordersAtPrice
+				break
+			}
+		}
+
+	case msgjson.SellOrderNum:
+		ordersAtPrice, ok := ob.sells[order.Rate]
+		if !ok {
+			return fmt.Errorf("no sell orders found at price %v", order.Rate)
+		}
+
+		for idx := len(ordersAtPrice) - 1; idx >= 0; idx-- {
+			if bytes.Equal(ordersAtPrice[idx].OrderID, order.OrderID) {
+				ordersAtPrice = append(ordersAtPrice[:idx], ordersAtPrice[idx+1:]...)
+				ob.sells[order.Rate] = ordersAtPrice
+				break
+			}
+		}
+	}
+
+	delete(ob.orders, order.OrderID.String())
+
+	return nil
+}
+
+// Initialize synchronizes a client tracked order book with an order
+// book snapshot.
+func (ob *OrderBook) Initialize(snapshot *msgjson.OrderBook) error {
+	ob.mtx.Lock()
+	defer ob.mtx.Unlock()
+
+	if ob.synced {
+		return fmt.Errorf("reinitializing a synced order book")
+	}
+
+	ob.orders = make(map[string]*msgjson.BookOrderNote, len(snapshot.Orders))
+	ob.buys = make(map[uint64][]*msgjson.BookOrderNote)
+	ob.sells = make(map[uint64][]*msgjson.BookOrderNote)
+
+	for _, order := range snapshot.Orders {
+		ob.addOrder(order)
+	}
+
+	ob.seq = snapshot.Seq
+	ob.marketID = snapshot.MarketID
+	ob.synced = true
+
+	return nil
+}
+
+// Book adds a new order to the order book.
+func (ob *OrderBook) Book(order *msgjson.BookOrderNote) error {
+	ob.mtx.Lock()
+	defer ob.mtx.Unlock()
+
+	if !ob.synced {
+		return fmt.Errorf("order book is unsynced")
+	}
+
+	if ob.marketID != order.MarketID {
+		return fmt.Errorf("market id mismatch, expected %s got %s",
+			ob.marketID, order.MarketID)
+	}
+
+	expectedSeq := ob.seq + 1
+	if expectedSeq != order.Seq {
+		ob.synced = false
+		return fmt.Errorf("order book out of sync, expected %d got %d",
+			expectedSeq, order.Seq)
+	}
+
+	ob.seq++
+	ob.addOrder(order)
+	return nil
+}
+
+// Unbook removes an order from the order book.
+func (ob *OrderBook) Unbook(unbookOrder *msgjson.UnbookOrderNote) error {
+	ob.mtx.Lock()
+	defer ob.mtx.Unlock()
+
+	if !ob.synced {
+		return fmt.Errorf("order book is unsynced")
+	}
+
+	if ob.marketID != unbookOrder.MarketID {
+		return fmt.Errorf("market id mismatch, expected %s got %s",
+			ob.marketID, unbookOrder.MarketID)
+	}
+
+	order, ok := ob.orders[unbookOrder.OrderID.String()]
+	if !ok {
+		return fmt.Errorf("no order found with id %v",
+			unbookOrder.OrderID.String())
+	}
```

Comment:
Also in this case, shouldn't `ob.seq` be incremented anyway since it was indeed for the correct market even though the order was not found?  BTW, what could it mean if the book is syncd but the order is not found?

Reply:
I'd even go as far as to say that any error like this should be treated as critical and the entire order book re-synced.